knowledge-base/summary-per-page-when-group-is-on-multiple-pages.md

```diff
@@ -0,0 +1,191 @@
+
+
+## Description
+For clarity, I will explain the requirement in a particular context, as requested by a user.  
+When generating a report that displays the list of movements for a list of accounts, it's possible (and quite frequent) for a single account to require multiple pages for 
+printing. In order to help the users understand what they are seeing and potentially quickly find errors in their external documents (for instance, bank statement), it is 
+convenient to print on the bottom of each page the total credit, debit amount up to this point in the printout. The above should only be displayed when a group is printed on
+multiple pages and not on the last page.
+
+Here is how this can be done with Telerik Reporting.
+
+## Solution
+Next, you may find two [custom user aggregate functions](../expressions-user-aggregate-functions) that can be used to accomplish the requirement. The first one is for the 
+[Page Header](../designing-reports-creating-page-headers-and-footers):
+```CSharp
+[AggregateFunction(Description = "Special sum aggregate. Output: (value1, value2, ...)", Name = "PageHeaderSumFromPrevPage")]
+class PageHeaderSumFromPrevPage : IAggregateFunction
+{
+    [ThreadStatic]
+    static decimal result;
+
+    [ThreadStatic]
+    static string currentGroupByValue;
+
+    [ThreadStatic]
+    static int currentPage = -1;
+
+    [ThreadStatic]
+    static List<object[]> currentPageValues;
+
+
+    public void Accumulate(object[] values)
+    {
+        var page = (int)values[2];
+        if (page != currentPage)
+        {
+            currentPage = page;
+
+            for (int index = 0; index < currentPageValues.Count; index++)
+            {
+                this.AccumulateCore(currentPageValues[index]);
+            }
+            currentPageValues.Clear();
+        }
+
+        currentPageValues.Add(values);
+    }
+
+    void AccumulateCore(object[] values)
+    {
+        // The aggregate function expects one parameter
+        object value = values[0];
+
+        // null values are not aggregated
+        if (null == value)
+        {
+            return;
+        }
+
+        var groupByValue = (string)values[1];
+
+        if (groupByValue != currentGroupByValue)
+        {
+            currentGroupByValue = groupByValue;
+            result = 0M;
+        }
+
+        result += (decimal)value;
+    }
+
+    public object GetValue()
+    {
+        if (currentPage == 1)
+        {
+            return 0;
+        }
+
+        return result;
+    }
+
+    public void Init()
+    {
+        if (currentPageValues == null)
+        {
+            currentPageValues = new List<object[]>();
+        }
+    }
+
+    public void Merge(IAggregateFunction aggregateFunction)
+    {
+    }
+}
+```
+The aggregate _PageHeaderSumFromPrevPage_ accumulates the corresponding 
+value (the first argument) inside the group (grouped by the second argument) up to the end of the previous page. The current page is provided as the third argument.
+
```

Comment:
e page footer aggregate is the simper one so I would start with it.
Explain that the fields need to be static, because for each page a new instance of the aggregate gets created and we need to preserve the data from all pages. On the other hand these are marked [ThreadStatic] so that if two report renderings occur simultaneously (like in web app) these do not interfere.

Reply:
And for the header aggregate explain that the idea derives from the other one, but we defer the accumulation of the current page's data needs to be postponed after the current page hence we need the page number to be passed as well.